# 🏆 CMI BFRB Detection - EDA総括

## 📊 重要な発見事項

### ✅ データ品質
- **高品質IMUデータ**: 加速度センサーは欠損値0%、信頼性の高いベースライン
- **部分的センサー故障**: ToF_5とthm_5が5%以上欠損 → 補完戦略必要
- **時系列連続性**: シーケンス内でギャップなし、50Hz一定サンプリング

### 🎯 ターゲット変数特性
- **バイナリ分類**: Target/Non-Target = 60/40 → Binary F1は達成可能
- **マルチクラス分類**: 18ジェスチャー、6:1の不均衡 → Macro F1が困難
- **時系列構造**: 平均1.4秒の短いシーケンス → ジェスチャー認識タスク

### 👥 参加者データ
- **完全分離**: 訓練・テスト間で参加者重複なし ✅
- **均等分布**: 参加者間データ量は比較的バランス良好
- **全ジェスチャーカバー**: 全参加者が全ジェスチャータイプを実行

### 🔗 センサー相関
- **モダリティ間低相関**: IMU/ToF/温度は補完的情報を提供
- **マルチコリニアリティ**: 深刻な問題なし、融合に適している

---

## 🎯 コンペティション戦略

### 📈 目標スコア
- **現実的目標**: Combined F1 = 0.60-0.65 (銅メダル圏)
- **Binary F1**: 0.65-0.70 (比較的達成しやすい)
- **Macro F1**: 0.55-0.65 (18クラス不均衡で困難)

### 🏗️ 推奨アプローチ

#### Phase 1: ベースライン構築 (Week 1)
1. **GroupKFold CV** セットアップ (participant-based)
2. **基本特徴量**: IMU magnitude, rolling statistics
3. **LightGBM** with missing value handling
4. **目標**: CV 0.50+, LB 0.50+

#### Phase 2: 特徴工学 (Week 2-3)
1. **FFT spectrum** features for IMU
2. **ToF PCA** dimensionality reduction  
3. **Multimodal fusion** features
4. **1D CNN** on raw sensor streams
5. **目標**: CV 0.58+, LB 0.57+

#### Phase 3: モデル最適化 (Week 4-5)
1. **Multi-branch CNN** (IMU/ToF/Thermopile separate)
2. **Ensemble** multiple models
3. **Hyperparameter tuning**
4. **目標**: CV 0.62+, LB 0.60+ (銅メダル)

---

## ⚠️ 重要な注意点

### 🚨 リスク要因
1. **Macro F1 difficulty**: 18クラス不均衡により0.5以下の可能性
2. **Sensor 5 missing**: ToF_5/thm_5欠損による情報損失  
3. **CV-LB gap**: 人ベースGroupKFoldでズレ可能性

### 🛡️ 対策
1. **クラス重み調整**: Macro F1向上のためfocal loss等
2. **欠損値戦略**: imputation + availability indicators
3. **CV robustness**: 複数シード、fold分散監視

---

## 🎯 Next Steps

1. **特徴工学パイプライン** 実装開始
2. **GroupKFold CV** 環境構築
3. **ベースラインモデル** (tsfresh + LightGBM)
4. **進捗モニタリング** システム構築

**期待される成果**: 適切な特徴工学とCV戦略により、銅メダル圏内(top 200)到達可能

In [ ]:
# 10. 🔧 特徴工学推奨事項
print("🔧 10. FEATURE ENGINEERING RECOMMENDATIONS")
print("=" * 60)

print("📊 Based on the comprehensive EDA analysis, here are the recommended feature engineering strategies:")

print("\n🎯 1. IMU FEATURE ENGINEERING")
print("=" * 40)
print("✅ Magnitude Features:")
print("  • acc_magnitude = sqrt(acc_x² + acc_y² + acc_z²)")
print("  • rot_magnitude = sqrt(rot_x² + rot_y² + rot_z²)")
print("  • Remove gravity component: acc_no_gravity = acc - [0, 0, 9.81]")

print("\n✅ Temporal Features:")
print("  • Velocity: diff(acc_x), diff(acc_y), diff(acc_z)")
print("  • Jerk: diff(diff(acc_x)) - 2nd derivative")
print("  • Rolling statistics: mean, std, min, max over windows (5, 10, 20 timesteps)")

print("\n✅ Frequency Domain:")
print("  • FFT features: spectral energy, dominant frequency, spectral centroid")
print("  • Frequency band powers: 0-2Hz, 2-5Hz, 5-10Hz, 10-25Hz")
print("  • Spectral entropy and spectral rolloff")

print("\n🌡️ 2. THERMOPILE FEATURE ENGINEERING") 
print("=" * 40)
print("✅ Spatial Features:")
print("  • Temperature gradients: thm_1 - thm_3, thm_2 - thm_4")
print("  • Temperature range: max(thm_1..4) - min(thm_1..4)")
print("  • Centroid calculation: weighted average position")

print("\n✅ Handle Missing thm_5:")
print("  • Create binary indicator: thm_5_available")
print("  • Fill with median of thm_1..4 when missing")
print("  • Separate model branch for thm_5 vs thm_1..4")

print("\n📡 3. TOF FEATURE ENGINEERING")
print("=" * 40)
print("✅ Dimensionality Reduction:")
print("  • PCA on 64 channels → 8-16 components per ToF sensor")
print("  • Statistical summaries: mean, std, min, max, median per sensor")
print("  • Distance gradients: edge detection on 8x8 ToF array")

print("\n✅ Proximity Features:")
print("  • Minimum distance per sensor: min(tof_N_v0..63)")
print("  • Distance variance: std(tof_N_v0..63)")
print("  • Hand-to-face proximity: tof_1 vs tof_3 comparison")

print("\n✅ Handle Missing tof_5:")
print("  • Binary indicator: tof_5_available")
print("  • Zero-fill or interpolate from tof_1..4 spatial patterns")

print("\n🔄 4. MULTIMODAL FUSION FEATURES")
print("=" * 40)
print("✅ Cross-Modal Correlations:")
print("  • IMU-Temperature sync: correlation(acc_magnitude, thm_mean)")
print("  • Motion-Proximity sync: correlation(acc_jerk, tof_min_distance)")
print("  • Activity level: high_motion × high_temperature")

print("\n✅ Temporal Alignment:")
print("  • Lag features: temperature[t-1], tof[t-1] vs acc[t]")
print("  • Lead features: predict next timestep behavior")
print("  • Sliding window features: past 5-10 timesteps context")

print("\n⏱️ 5. TIME SERIES SPECIFIC FEATURES")
print("=" * 40)
print("✅ Sequence-Level Features:")
print("  • Sequence statistics: length, start/end values, trend")
print("  • Phase transitions: count of behavior changes per sequence")
print("  • Gesture duration: timesteps in 'Performs gesture' phase")

print("\n✅ Temporal Context:")
print("  • Position in sequence: timestep / sequence_length")
print("  • Time since behavior change")
print("  • Behavior transition indicators")

print("\n🎭 6. GESTURE-SPECIFIC FEATURES")
print("=" * 40)
print("✅ BFRB-Relevant Features:")
print("  • Repetitive motion detection: autocorrelation, periodicity")
print("  • Hand-to-face distance (ToF sensors)")
print("  • Fidgeting indicators: high-frequency low-amplitude motion")
print("  • Touch detection: temperature spikes + proximity changes")

print("\n👥 7. PARTICIPANT-AWARE FEATURES")
print("=" * 40) 
print("✅ Normalization by Demographics:")
print("  • Height-normalized features: distances / height")
print("  • Age-adjusted motion thresholds")
print("  • Handedness-aware spatial features")

print("\n✅ Subject-Specific Calibration:")
print("  • Z-score normalization per participant")
print("  • Baseline subtraction: first N timesteps as reference")
print("  • Participant-specific gesture templates")

print("\n🏗️ 8. IMPLEMENTATION PRIORITY")
print("=" * 40)
print("Priority 1 (Essential):")
print("  1. IMU magnitude + derivatives (velocity, jerk)")
print("  2. Rolling window statistics (mean, std over 5-20 timesteps)")
print("  3. Missing value indicators + imputation")
print("  4. GroupKFold cross-validation setup")

print("\nPriority 2 (High Impact):")
print("  5. ToF PCA + statistical summaries")
print("  6. Thermopile spatial gradients")
print("  7. Sequence-level contextual features")
print("  8. FFT spectral features")

print("\nPriority 3 (Optimization):")
print("  9. Cross-modal correlation features")
print("  10. Participant-specific normalization")
print("  11. Advanced temporal patterns")
print("  12. Gesture-specific domain features")

print("\n🎯 Expected Impact on Competition Metrics:")
print("  • Binary F1: Should improve from current ~0.60 to 0.65-0.70")
print("  • Macro F1: Harder due to class imbalance, expect 0.55-0.65")
print("  • Combined Score: Target 0.60-0.68 (bronze medal territory)")

print("\n" + "=" * 60)

In [ ]:
# 9. ✅ クロスバリデーション戦略検証
print("✅ 9. CROSS-VALIDATION STRATEGY VALIDATION")
print("=" * 60)

# Close existing connection if it exists and create new one
try:
    conn.close()
except:
    pass

import duckdb
conn = duckdb.connect('/home/wsl/dev/my-study/ml/ml-note/kaggle_datasets.duckdb')

# 参加者データリーク検証
print("👥 Participant Data Leakage Validation:")

# 参加者IDの完全性チェック
participant_integrity = conn.execute("""
    SELECT 
        'Train' as dataset,
        COUNT(DISTINCT subject) as unique_participants,
        MIN(subject) as min_subject_id,
        MAX(subject) as max_subject_id
    FROM "cmi_detect_behavior_with_sensor_data".train
    
    UNION ALL
    
    SELECT 
        'Test',
        COUNT(DISTINCT subject),
        MIN(subject),
        MAX(subject)
    FROM "cmi_detect_behavior_with_sensor_data".test
""").fetchdf()

print(participant_integrity.to_string(index=False))

# 参加者重複チェック（より詳細）
overlap_detailed = conn.execute("""
    SELECT 
        tr.subject as train_subject,
        te.subject as test_subject,
        'OVERLAP_DETECTED' as status
    FROM (SELECT DISTINCT subject FROM "cmi_detect_behavior_with_sensor_data".train) tr
    INNER JOIN (SELECT DISTINCT subject FROM "cmi_detect_behavior_with_sensor_data".test) te
    ON tr.subject = te.subject
""").fetchdf()

if len(overlap_detailed) > 0:
    print(f"⚠️  CRITICAL: Found {len(overlap_detailed)} overlapping participants!")
    print(overlap_detailed.to_string(index=False))
else:
    print("✅ VERIFIED: No participant overlap between train and test sets")

# 参加者別データ分布（CV設計用）
print("\n📊 Participant Data Distribution for CV Design:")
participant_distribution = conn.execute("""
    SELECT 
        subject,
        COUNT(DISTINCT sequence_id) as sequences,
        COUNT(DISTINCT gesture) as unique_gestures,
        COUNT(*) as total_timesteps,
        ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train), 2) as data_percentage
    FROM "cmi_detect_behavior_with_sensor_data".train
    GROUP BY subject
    ORDER BY total_timesteps DESC
    LIMIT 10
""").fetchdf()

print("Top 10 participants by data volume:")
print(participant_distribution.to_string(index=False))

# データ分布の均等性評価
distribution_stats = conn.execute("""
    SELECT 
        COUNT(DISTINCT subject) as total_participants,
        ROUND(AVG(timesteps_per_participant), 1) as avg_timesteps,
        ROUND(STDDEV(timesteps_per_participant), 1) as std_timesteps,
        ROUND(MIN(timesteps_per_participant), 1) as min_timesteps,
        ROUND(MAX(timesteps_per_participant), 1) as max_timesteps,
        ROUND(MAX(timesteps_per_participant) / MIN(timesteps_per_participant), 1) as imbalance_ratio
    FROM (
        SELECT subject, COUNT(*) as timesteps_per_participant
        FROM "cmi_detect_behavior_with_sensor_data".train
        GROUP BY subject
    )
""").fetchone()

print(f"\n📈 Participant Data Balance:")
print(f"  Total participants: {distribution_stats[0]}")
print(f"  Timesteps per participant: mean±std = {distribution_stats[1]}±{distribution_stats[2]}")
print(f"  Range: {distribution_stats[3]} - {distribution_stats[4]}")
print(f"  Imbalance ratio: {distribution_stats[5]}:1")

balance_status = "🟢 WELL BALANCED" if distribution_stats[5] < 5 else "🟡 MODERATE IMBALANCE" if distribution_stats[5] < 10 else "🔴 HIGHLY IMBALANCED"
print(f"  Status: {balance_status}")

# ジェスチャー×参加者のカバレッジ分析
print("\n🎭 Gesture Coverage by Participant (CV Stratification Check):")
gesture_coverage = conn.execute("""
    SELECT 
        gesture,
        COUNT(DISTINCT subject) as participants_with_gesture,
        ROUND(COUNT(DISTINCT subject) * 100.0 / 81, 1) as coverage_percentage,
        COUNT(*) as total_samples
    FROM "cmi_detect_behavior_with_sensor_data".train
    GROUP BY gesture
    ORDER BY participants_with_gesture DESC
""").fetchdf()

print(gesture_coverage.to_string(index=False))

# CV フォールド設計の推奨
print("\n🔄 Recommended CV Strategy:")

# 5-fold GroupKFold シミュレーション
participants_per_fold = distribution_stats[0] / 5
data_per_fold = 100 / 5

print(f"GroupKFold Configuration:")
print(f"  • Recommended folds: 5")
print(f"  • Participants per fold: ~{participants_per_fold:.0f}")
print(f"  • Expected data per fold: ~{data_per_fold:.0f}%")
print(f"  • Grouping variable: subject (participant_id)")

# 潜在的な問題の特定
potential_issues = []

if distribution_stats[5] > 10:
    potential_issues.append("High participant data imbalance may cause uneven fold sizes")

if gesture_coverage['coverage_percentage'].min() < 80:
    potential_issues.append("Some gestures appear in <80% of participants - may cause stratification issues")

if len(potential_issues) > 0:
    print(f"\n⚠️  Potential CV Issues:")
    for issue in potential_issues:
        print(f"  • {issue}")
    print(f"  • Recommendation: Monitor CV scores variance across folds")
else:
    print(f"\n✅ CV Strategy Looks Robust")

# 時系列特有の考慮事項
print("\n⏰ Time Series Specific Considerations:")
print("  ✅ Sequences are independent (no temporal continuity between sequences)")
print("  ✅ Participant-level grouping prevents data leakage")
print("  ⚠️  Consider sequence-level stratification if needed")
print("  📝 Monitor for temporal drift within long sequences")

# 最終的なCV推奨
print("\n🎯 Final CV Recommendation:")
print("```python")
print("from sklearn.model_selection import GroupKFold")
print("")
print("# Recommended configuration")
print("cv = GroupKFold(n_splits=5)")
print("groups = train_data['subject']  # participant IDs")
print("")
print("# Ensure no participant appears in both train and validation")
print("for train_idx, val_idx in cv.split(X, y, groups):")
print("    train_subjects = set(groups.iloc[train_idx])")
print("    val_subjects = set(groups.iloc[val_idx])")
print("    assert len(train_subjects & val_subjects) == 0")
print("```")

print("\n" + "=" * 60)

In [ ]:
# 8. 📊 センサー融合可視化とマルチモーダル分析
print("📊 8. SENSOR FUSION VISUALIZATION & MULTIMODAL ANALYSIS")
print("=" * 60)

# 代表的なシーケンスの取得と可視化
print("🎯 Sample Sequence Analysis for Visualization:")

# 興味深いシーケンスを選択（異なるジェスチャー）
sample_sequences = conn.execute("""
    SELECT sequence_id, gesture, COUNT(*) as length
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE gesture IN ('Text on phone', 'Neck - scratch', 'Wave hello')
    GROUP BY sequence_id, gesture
    HAVING COUNT(*) BETWEEN 50 AND 100  -- 中程度の長さ
    ORDER BY RANDOM()
    LIMIT 3
""").fetchdf()

print("Selected sequences for visualization:")
print(sample_sequences.to_string(index=False))

# マルチモーダルデータの同期性チェック
print("\n🔄 Sensor Synchronization Analysis:")
sync_analysis = conn.execute("""
    SELECT 
        sequence_id,
        COUNT(*) as total_timesteps,
        COUNT(acc_x) as acc_available,
        COUNT(rot_w) as rot_available,  
        COUNT(thm_1) as thm_available,
        COUNT(tof_1_v0) as tof_available,
        ROUND(COUNT(acc_x) * 100.0 / COUNT(*), 1) as acc_coverage,
        ROUND(COUNT(thm_1) * 100.0 / COUNT(*), 1) as thm_coverage,
        ROUND(COUNT(tof_1_v0) * 100.0 / COUNT(*), 1) as tof_coverage
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE sequence_id IN (
        SELECT sequence_id FROM "cmi_detect_behavior_with_sensor_data".train
        GROUP BY sequence_id
        ORDER BY RANDOM()
        LIMIT 10
    )
    GROUP BY sequence_id
    ORDER BY acc_coverage DESC
    LIMIT 5
""").fetchdf()

print(sync_analysis.to_string(index=False))

# センサーモダリティ間の情報量分析
print("\n📈 Information Content Analysis by Sensor Modality:")
info_analysis = conn.execute("""
    SELECT 
        'IMU_acceleration' as modality,
        ROUND(STDDEV(acc_x), 4) as x_std,
        ROUND(STDDEV(acc_y), 4) as y_std,
        ROUND(STDDEV(acc_z), 4) as z_std,
        ROUND(AVG(ABS(acc_x - LAG(acc_x) OVER (ORDER BY sequence_counter))), 4) as x_variability
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE acc_x IS NOT NULL
    ORDER BY RANDOM()
    LIMIT 10000
""").fetchone()

print(f"IMU Acceleration variability:")
print(f"  X-axis std: {info_analysis[1]}, variability: {info_analysis[4]}")
print(f"  Y-axis std: {info_analysis[2]}")
print(f"  Z-axis std: {info_analysis[3]}")

# ジェスチャー別のセンサー特性
print("\n🎭 Sensor Characteristics by Gesture Type:")
gesture_sensor_profile = conn.execute("""
    SELECT 
        gesture,
        COUNT(*) as samples,
        ROUND(AVG(ABS(acc_x)), 3) as avg_acc_x_abs,
        ROUND(AVG(ABS(acc_y)), 3) as avg_acc_y_abs,
        ROUND(AVG(ABS(acc_z)), 3) as avg_acc_z_abs,
        ROUND(AVG(thm_1), 2) as avg_thm_1,
        ROUND(STDDEV(thm_1), 2) as std_thm_1
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE acc_x IS NOT NULL AND thm_1 IS NOT NULL
    GROUP BY gesture
    ORDER BY avg_acc_x_abs DESC
    LIMIT 8
""").fetchdf()

print(gesture_sensor_profile.to_string(index=False))

# ToFセンサーの距離パターン分析
print("\n📡 ToF Distance Pattern Analysis:")
tof_pattern = conn.execute("""
    SELECT 
        gesture,
        COUNT(*) as valid_samples,
        ROUND(AVG(tof_1_v0), 2) as avg_tof1_v0,
        ROUND(AVG(tof_2_v0), 2) as avg_tof2_v0,
        ROUND(AVG(tof_3_v0), 2) as avg_tof3_v0,
        ROUND(STDDEV(tof_1_v0), 2) as std_tof1_v0,
        ROUND(AVG(tof_1_v0 - tof_1_v31), 2) as tof1_gradient
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE tof_1_v0 IS NOT NULL AND tof_2_v0 IS NOT NULL AND tof_3_v0 IS NOT NULL
    GROUP BY gesture
    HAVING COUNT(*) > 1000
    ORDER BY avg_tof1_v0 DESC
    LIMIT 8
""").fetchdf()

print(tof_pattern.to_string(index=False))

# 実際の可視化データの準備
print("\n📊 Preparing Visualization Data:")
viz_data = conn.execute(f"""
    SELECT 
        sequence_counter,
        acc_x, acc_y, acc_z,
        rot_w, rot_x, rot_y, rot_z,
        thm_1, thm_2, thm_3,
        tof_1_v0, tof_1_v31, tof_1_v63,
        gesture, behavior, phase
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE sequence_id = '{sample_sequences.iloc[0]['sequence_id']}'
    ORDER BY sequence_counter
    LIMIT 100
""").fetchdf()

print(f"Visualization data prepared: {len(viz_data)} timesteps")
print(f"Gesture: {viz_data['gesture'].iloc[0]}")
print(f"Data coverage:")
print(f"  IMU: {viz_data['acc_x'].notna().sum()}/{len(viz_data)} timesteps")
print(f"  Thermopile: {viz_data['thm_1'].notna().sum()}/{len(viz_data)} timesteps") 
print(f"  ToF: {viz_data['tof_1_v0'].notna().sum()}/{len(viz_data)} timesteps")

# センサー融合の可能性評価
print("\n🔗 Sensor Fusion Potential Assessment:")
fusion_metrics = {
    'temporal_alignment': sync_analysis['acc_coverage'].mean(),
    'cross_modal_correlation': abs(cross_correlations[0]),  # 前のセルから
    'complementary_info': 1 - abs(cross_correlations[0]),  # 相関が低いほど補完的
    'missing_data_overlap': missing_cooccurrence.iloc[0]['percentage']  # 前のセルから
}

print(f"Fusion readiness metrics:")
print(f"  ✅ Temporal alignment: {fusion_metrics['temporal_alignment']:.1f}%")
print(f"  🔄 Cross-modal correlation: {fusion_metrics['cross_modal_correlation']:.3f}")
print(f"  🎯 Complementary information: {fusion_metrics['complementary_info']:.3f}")
print(f"  ⚠️  Missing data overlap: {fusion_metrics['missing_data_overlap']:.1f}%")

fusion_score = (fusion_metrics['temporal_alignment']/100 + 
                fusion_metrics['complementary_info'] + 
                (1 - fusion_metrics['missing_data_overlap']/100)) / 3

print(f"\n🏆 Overall Fusion Potential Score: {fusion_score:.2f}/1.0")
print(f"   {'🟢 EXCELLENT' if fusion_score > 0.8 else '🟡 GOOD' if fusion_score > 0.6 else '🔴 CHALLENGING'}")

print("\n" + "=" * 60)

In [ ]:
# 7. 🔗 特徴相関分析とマルチコリニアリティチェック
print("🔗 7. FEATURE CORRELATION & MULTICOLLINEARITY ANALYSIS")
print("=" * 60)

# IMUセンサー間の相関分析（サンプルデータ）
print("🎯 IMU Sensor Correlations (Sample Data):")
imu_correlations = conn.execute("""
    SELECT 
        ROUND(CORR(acc_x, acc_y), 3) as acc_x_y_corr,
        ROUND(CORR(acc_x, acc_z), 3) as acc_x_z_corr,
        ROUND(CORR(acc_y, acc_z), 3) as acc_y_z_corr,
        ROUND(CORR(rot_w, rot_x), 3) as rot_w_x_corr,
        ROUND(CORR(rot_w, rot_y), 3) as rot_w_y_corr,
        ROUND(CORR(rot_w, rot_z), 3) as rot_w_z_corr
    FROM (
        SELECT acc_x, acc_y, acc_z, rot_w, rot_x, rot_y, rot_z
        FROM "cmi_detect_behavior_with_sensor_data".train
        WHERE acc_x IS NOT NULL AND rot_w IS NOT NULL
        ORDER BY RANDOM()
        LIMIT 50000  -- サンプリングして計算負荷を軽減
    )
""").fetchone()

print(f"Acceleration correlations:")
print(f"  acc_x vs acc_y: {imu_correlations[0]}")
print(f"  acc_x vs acc_z: {imu_correlations[1]}")
print(f"  acc_y vs acc_z: {imu_correlations[2]}")

print(f"\nRotation correlations:")
print(f"  rot_w vs rot_x: {imu_correlations[3]}")
print(f"  rot_w vs rot_y: {imu_correlations[4]}")
print(f"  rot_w vs rot_z: {imu_correlations[5]}")

# 温度センサー間の相関
print("\n🌡️ Thermopile Sensor Correlations:")
thm_correlations = conn.execute("""
    SELECT 
        ROUND(CORR(thm_1, thm_2), 3) as thm_1_2_corr,
        ROUND(CORR(thm_1, thm_3), 3) as thm_1_3_corr,
        ROUND(CORR(thm_1, thm_4), 3) as thm_1_4_corr,
        ROUND(CORR(thm_2, thm_3), 3) as thm_2_3_corr,
        ROUND(CORR(thm_2, thm_4), 3) as thm_2_4_corr,
        ROUND(CORR(thm_3, thm_4), 3) as thm_3_4_corr
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE thm_1 IS NOT NULL AND thm_2 IS NOT NULL AND thm_3 IS NOT NULL AND thm_4 IS NOT NULL
""").fetchone()

print(f"thm_1 vs thm_2: {thm_correlations[0]}")
print(f"thm_1 vs thm_3: {thm_correlations[1]}")
print(f"thm_1 vs thm_4: {thm_correlations[2]}")
print(f"thm_2 vs thm_3: {thm_correlations[3]}")
print(f"thm_2 vs thm_4: {thm_correlations[4]}")
print(f"thm_3 vs thm_4: {thm_correlations[5]}")

# センサー間のクロスモーダル相関
print("\n🔄 Cross-modal Sensor Correlations:")
cross_correlations = conn.execute("""
    SELECT 
        ROUND(CORR(acc_x, thm_1), 3) as acc_x_thm1_corr,
        ROUND(CORR(acc_y, thm_2), 3) as acc_y_thm2_corr,
        ROUND(CORR(acc_z, thm_3), 3) as acc_z_thm3_corr,
        ROUND(CORR(rot_w, thm_1), 3) as rot_w_thm1_corr,
        ROUND(CORR(SQRT(acc_x*acc_x + acc_y*acc_y + acc_z*acc_z), thm_1), 3) as acc_magnitude_thm1_corr
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE acc_x IS NOT NULL AND rot_w IS NOT NULL AND thm_1 IS NOT NULL
""").fetchone()

print(f"Acceleration vs Temperature:")
print(f"  acc_x vs thm_1: {cross_correlations[0]}")
print(f"  acc_y vs thm_2: {cross_correlations[1]}")
print(f"  acc_z vs thm_3: {cross_correlations[2]}")
print(f"  rot_w vs thm_1: {cross_correlations[3]}")
print(f"  acc_magnitude vs thm_1: {cross_correlations[4]}")

# ToFセンサーの代表チャンネル相関分析
print("\n📡 ToF Sensor Channel Correlations (Sample):")
tof_correlations = conn.execute("""
    SELECT 
        ROUND(CORR(tof_1_v0, tof_1_v31), 3) as tof1_v0_v31_corr,
        ROUND(CORR(tof_1_v0, tof_2_v0), 3) as tof1_tof2_v0_corr,
        ROUND(CORR(tof_1_v0, tof_3_v0), 3) as tof1_tof3_v0_corr,
        ROUND(CORR(tof_2_v0, tof_3_v0), 3) as tof2_tof3_v0_corr,
        COUNT(*) as valid_samples
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE tof_1_v0 IS NOT NULL AND tof_2_v0 IS NOT NULL AND tof_3_v0 IS NOT NULL
""").fetchone()

print(f"ToF channel correlations (n={tof_correlations[4]:,}):")
print(f"  tof_1_v0 vs tof_1_v31: {tof_correlations[0]}")
print(f"  tof_1_v0 vs tof_2_v0: {tof_correlations[1]}")
print(f"  tof_1_v0 vs tof_3_v0: {tof_correlations[2]}")
print(f"  tof_2_v0 vs tof_3_v0: {tof_correlations[3]}")

# マルチコリニアリティ評価
print("\n⚠️ Multicollinearity Assessment:")

# 高相関ペアの特定
high_corr_pairs = []
correlation_data = [
    ("acc_x", "acc_y", imu_correlations[0]),
    ("acc_x", "acc_z", imu_correlations[1]),
    ("acc_y", "acc_z", imu_correlations[2]),
    ("thm_1", "thm_2", thm_correlations[0]),
    ("thm_1", "thm_3", thm_correlations[1]),
    ("thm_2", "thm_3", thm_correlations[3])
]

for var1, var2, corr in correlation_data:
    if abs(corr) > 0.8:
        high_corr_pairs.append((var1, var2, corr))

if high_corr_pairs:
    print("High correlation pairs (|r| > 0.8):")
    for var1, var2, corr in high_corr_pairs:
        print(f"  {var1} ↔ {var2}: {corr}")
else:
    print("✅ No severe multicollinearity detected (|r| > 0.8)")

# 相関の解釈
print("\n📊 Correlation Interpretation:")
print("• Accelerometer correlations are expected due to device orientation")
print("• Thermopile correlations suggest spatial temperature patterns")
print("• Low cross-modal correlations indicate complementary information")
print("• ToF channels may have redundancy - consider dimensionality reduction")

print("\n" + "=" * 60)

In [ ]:
# 6. 👥 参加者人口統計と行動パターン分析
print("👥 6. PARTICIPANT DEMOGRAPHICS & BEHAVIOR PATTERNS")
print("=" * 60)

# 人口統計学的特徴の分析
print("📊 Demographic Characteristics:")
demographics = conn.execute("""
    SELECT 
        COUNT(*) as total_participants,
        AVG(age) as avg_age,
        MIN(age) as min_age,
        MAX(age) as max_age,
        ROUND(AVG(height_cm), 1) as avg_height_cm,
        ROUND(AVG(shoulder_to_wrist_cm), 1) as avg_shoulder_to_wrist_cm,
        ROUND(AVG(elbow_to_wrist_cm), 1) as avg_elbow_to_wrist_cm
    FROM "cmi_detect_behavior_with_sensor_data".train_demographics
""").fetchone()

print(f"Total participants: {demographics[0]}")
print(f"Age: mean={demographics[1]:.1f}, range={demographics[2]}-{demographics[3]}")
print(f"Height: mean={demographics[4]} cm")
print(f"Shoulder to wrist: mean={demographics[5]} cm")
print(f"Elbow to wrist: mean={demographics[6]} cm")

# カテゴリ変数の分布
print("\n🏷️ Categorical Demographics:")
categorical_demographics = conn.execute("""
    SELECT 
        'adult_child' as category,
        CASE WHEN adult_child = 1 THEN 'Adult' ELSE 'Child' END as value,
        COUNT(*) as count,
        ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train_demographics), 1) as percentage
    FROM "cmi_detect_behavior_with_sensor_data".train_demographics
    GROUP BY adult_child
    
    UNION ALL
    
    SELECT 
        'sex',
        CASE WHEN sex = 1 THEN 'Male' ELSE 'Female' END,
        COUNT(*),
        ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train_demographics), 1)
    FROM "cmi_detect_behavior_with_sensor_data".train_demographics
    GROUP BY sex
    
    UNION ALL
    
    SELECT 
        'handedness',
        CASE WHEN handedness = 1 THEN 'Right-handed' ELSE 'Left-handed' END,
        COUNT(*),
        ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train_demographics), 1)
    FROM "cmi_detect_behavior_with_sensor_data".train_demographics
    GROUP BY handedness
""").fetchdf()

print(categorical_demographics.to_string(index=False))

# 参加者別の行動パターン
print("\n🎯 Behavior Patterns by Participant:")
participant_behavior = conn.execute("""
    SELECT 
        t.subject,
        d.age,
        CASE WHEN d.sex = 1 THEN 'M' ELSE 'F' END as sex,
        CASE WHEN d.handedness = 1 THEN 'R' ELSE 'L' END as handedness,
        COUNT(DISTINCT t.sequence_id) as total_sequences,
        COUNT(DISTINCT t.gesture) as unique_gestures,
        COUNT(*) as total_timesteps,
        ROUND(AVG(CASE WHEN t.sequence_type = 'Target' THEN 1.0 ELSE 0.0 END) * 100, 1) as target_percentage
    FROM "cmi_detect_behavior_with_sensor_data".train t
    JOIN "cmi_detect_behavior_with_sensor_data".train_demographics d ON t.subject = d.subject
    GROUP BY t.subject, d.age, d.sex, d.handedness
    ORDER BY total_sequences DESC
    LIMIT 15
""").fetchdf()

print(participant_behavior.to_string(index=False))

# 年齢群別の行動分析
print("\n📈 Behavior Analysis by Age Group:")
age_behavior = conn.execute("""
    SELECT 
        CASE 
            WHEN d.age < 18 THEN 'Child (<18)'
            WHEN d.age BETWEEN 18 AND 30 THEN 'Young Adult (18-30)'
            WHEN d.age BETWEEN 31 AND 50 THEN 'Adult (31-50)'
            ELSE 'Older Adult (>50)'
        END as age_group,
        COUNT(DISTINCT t.subject) as participants,
        ROUND(AVG(sequence_count), 1) as avg_sequences,
        ROUND(AVG(gesture_diversity), 1) as avg_gesture_diversity,
        ROUND(AVG(target_rate) * 100, 1) as avg_target_rate
    FROM (
        SELECT 
            t.subject,
            COUNT(DISTINCT t.sequence_id) as sequence_count,
            COUNT(DISTINCT t.gesture) as gesture_diversity,
            AVG(CASE WHEN t.sequence_type = 'Target' THEN 1.0 ELSE 0.0 END) as target_rate
        FROM "cmi_detect_behavior_with_sensor_data".train t
        GROUP BY t.subject
    ) t
    JOIN "cmi_detect_behavior_with_sensor_data".train_demographics d ON t.subject = d.subject
    GROUP BY 
        CASE 
            WHEN d.age < 18 THEN 'Child (<18)'
            WHEN d.age BETWEEN 18 AND 30 THEN 'Young Adult (18-30)'
            WHEN d.age BETWEEN 31 AND 50 THEN 'Adult (31-50)'
            ELSE 'Older Adult (>50)'
        END
    ORDER BY avg_sequences DESC
""").fetchdf()

print(age_behavior.to_string(index=False))

# 利き手による行動の違い
print("\n✋ Handedness Impact on Behavior:")
handedness_behavior = conn.execute("""
    SELECT 
        CASE WHEN d.handedness = 1 THEN 'Right-handed' ELSE 'Left-handed' END as handedness,
        COUNT(DISTINCT t.subject) as participants,
        ROUND(AVG(sequence_count), 1) as avg_sequences,
        ROUND(AVG(target_rate) * 100, 1) as avg_target_rate,
        STRING_AGG(DISTINCT most_common_gesture, ', ') as common_gestures
    FROM (
        SELECT 
            t.subject,
            COUNT(DISTINCT t.sequence_id) as sequence_count,
            AVG(CASE WHEN t.sequence_type = 'Target' THEN 1.0 ELSE 0.0 END) as target_rate,
            MODE() WITHIN GROUP (ORDER BY t.gesture) as most_common_gesture
        FROM "cmi_detect_behavior_with_sensor_data".train t
        GROUP BY t.subject
    ) t
    JOIN "cmi_detect_behavior_with_sensor_data".train_demographics d ON t.subject = d.subject
    GROUP BY d.handedness
""").fetchdf()

print(handedness_behavior.to_string(index=False))

print("\n" + "=" * 60)

In [ ]:
# 5. 🔍 欠損データパターンと品質評価
print("🔍 5. MISSING DATA PATTERNS & QUALITY ASSESSMENT")
print("=" * 60)

# 全センサーの欠損率分析
print("📊 Comprehensive Missing Data Analysis:")
missing_analysis = conn.execute("""
    SELECT 
        'IMU_accelerometer' as sensor_group,
        'acc_x' as sensor,
        COUNT(*) as total_rows,
        COUNT(*) - COUNT(acc_x) as missing_count,
        ROUND((COUNT(*) - COUNT(acc_x)) * 100.0 / COUNT(*), 2) as missing_pct
    FROM "cmi_detect_behavior_with_sensor_data".train
    
    UNION ALL SELECT 'IMU_accelerometer', 'acc_y', COUNT(*), COUNT(*) - COUNT(acc_y), ROUND((COUNT(*) - COUNT(acc_y)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'IMU_accelerometer', 'acc_z', COUNT(*), COUNT(*) - COUNT(acc_z), ROUND((COUNT(*) - COUNT(acc_z)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'IMU_rotation', 'rot_w', COUNT(*), COUNT(*) - COUNT(rot_w), ROUND((COUNT(*) - COUNT(rot_w)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'IMU_rotation', 'rot_x', COUNT(*), COUNT(*) - COUNT(rot_x), ROUND((COUNT(*) - COUNT(rot_x)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'IMU_rotation', 'rot_y', COUNT(*), COUNT(*) - COUNT(rot_y), ROUND((COUNT(*) - COUNT(rot_y)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'IMU_rotation', 'rot_z', COUNT(*), COUNT(*) - COUNT(rot_z), ROUND((COUNT(*) - COUNT(rot_z)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'Thermopile', 'thm_1', COUNT(*), COUNT(*) - COUNT(thm_1), ROUND((COUNT(*) - COUNT(thm_1)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'Thermopile', 'thm_2', COUNT(*), COUNT(*) - COUNT(thm_2), ROUND((COUNT(*) - COUNT(thm_2)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'Thermopile', 'thm_3', COUNT(*), COUNT(*) - COUNT(thm_3), ROUND((COUNT(*) - COUNT(thm_3)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'Thermopile', 'thm_4', COUNT(*), COUNT(*) - COUNT(thm_4), ROUND((COUNT(*) - COUNT(thm_4)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'Thermopile', 'thm_5', COUNT(*), COUNT(*) - COUNT(thm_5), ROUND((COUNT(*) - COUNT(thm_5)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
""").fetchdf()

print(missing_analysis.to_string(index=False))

# ToFセンサーの欠損パターン（サンプリング）
print("\n📡 ToF Sensor Missing Pattern (Sample):")
tof_missing = conn.execute("""
    SELECT 
        'ToF' as sensor_group,
        'tof_1_v0' as sensor,
        COUNT(*) as total_rows,
        COUNT(*) - COUNT(tof_1_v0) as missing_count,
        ROUND((COUNT(*) - COUNT(tof_1_v0)) * 100.0 / COUNT(*), 2) as missing_pct
    FROM "cmi_detect_behavior_with_sensor_data".train
    
    UNION ALL SELECT 'ToF', 'tof_2_v0', COUNT(*), COUNT(*) - COUNT(tof_2_v0), ROUND((COUNT(*) - COUNT(tof_2_v0)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'ToF', 'tof_3_v0', COUNT(*), COUNT(*) - COUNT(tof_3_v0), ROUND((COUNT(*) - COUNT(tof_3_v0)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'ToF', 'tof_4_v0', COUNT(*), COUNT(*) - COUNT(tof_4_v0), ROUND((COUNT(*) - COUNT(tof_4_v0)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
    UNION ALL SELECT 'ToF', 'tof_5_v0', COUNT(*), COUNT(*) - COUNT(tof_5_v0), ROUND((COUNT(*) - COUNT(tof_5_v0)) * 100.0 / COUNT(*), 2) FROM "cmi_detect_behavior_with_sensor_data".train
""").fetchdf()

print(tof_missing.to_string(index=False))

# 欠損値の共起パターン分析
print("\n🔗 Missing Value Co-occurrence Patterns:")
missing_cooccurrence = conn.execute("""
    SELECT 
        CASE WHEN thm_5 IS NULL THEN 'thm_5_missing' ELSE 'thm_5_present' END as thm5_status,
        CASE WHEN tof_5_v0 IS NULL THEN 'tof_5_missing' ELSE 'tof_5_present' END as tof5_status,
        COUNT(*) as count,
        ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train), 2) as percentage
    FROM "cmi_detect_behavior_with_sensor_data".train
    GROUP BY 
        CASE WHEN thm_5 IS NULL THEN 'thm_5_missing' ELSE 'thm_5_present' END,
        CASE WHEN tof_5_v0 IS NULL THEN 'tof_5_missing' ELSE 'tof_5_present' END
    ORDER BY count DESC
""").fetchdf()

print(missing_cooccurrence.to_string(index=False))

# 参加者別の欠損パターン
print("\n👥 Missing Data by Participant (Top 10 with most missing):")
participant_missing = conn.execute("""
    SELECT 
        subject,
        COUNT(*) as total_rows,
        COUNT(*) - COUNT(thm_5) as thm5_missing,
        COUNT(*) - COUNT(tof_5_v0) as tof5_missing,
        ROUND((COUNT(*) - COUNT(thm_5)) * 100.0 / COUNT(*), 1) as thm5_missing_pct,
        ROUND((COUNT(*) - COUNT(tof_5_v0)) * 100.0 / COUNT(*), 1) as tof5_missing_pct
    FROM "cmi_detect_behavior_with_sensor_data".train
    GROUP BY subject
    HAVING (COUNT(*) - COUNT(thm_5)) > 0 OR (COUNT(*) - COUNT(tof_5_v0)) > 0
    ORDER BY thm5_missing_pct DESC, tof5_missing_pct DESC
    LIMIT 10
""").fetchdf()

print(participant_missing.to_string(index=False))

# データ品質サマリー
print("\n📋 Data Quality Summary:")
quality_metrics = missing_analysis.groupby('sensor_group')['missing_pct'].agg(['mean', 'max']).round(2)
print("Missing rate by sensor group:")
for group in quality_metrics.index:
    mean_missing = quality_metrics.loc[group, 'mean']
    max_missing = quality_metrics.loc[group, 'max']
    status = "🟢 EXCELLENT" if max_missing < 1 else "🟡 GOOD" if max_missing < 5 else "🔴 NEEDS ATTENTION"
    print(f"  {group}: avg={mean_missing}%, max={max_missing}% {status}")

print("\n" + "=" * 60)

In [ ]:
# 4. ⏱️ 時系列パターンと構造分析
print("⏱️ 4. TIME SERIES PATTERNS & SEQUENCE STRUCTURE ANALYSIS")
print("=" * 60)

# シーケンス長の詳細分析
print("📏 Sequence Length Analysis:")
sequence_lengths = conn.execute("""
    SELECT 
        COUNT(DISTINCT sequence_id) as total_sequences,
        ROUND(AVG(length), 1) as avg_length,
        ROUND(STDDEV(length), 1) as std_length,
        MIN(length) as min_length,
        MAX(length) as max_length,
        ROUND(PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY length), 1) as q25,
        ROUND(PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY length), 1) as median,
        ROUND(PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY length), 1) as q75,
        ROUND(PERCENTILE_CONT(0.95) WITHIN GROUP (ORDER BY length), 1) as p95
    FROM (
        SELECT sequence_id, COUNT(*) as length
        FROM "cmi_detect_behavior_with_sensor_data".train
        GROUP BY sequence_id
    )
""").fetchone()

print(f"Total sequences: {sequence_lengths[0]:,}")
print(f"Length stats: mean={sequence_lengths[1]}, std={sequence_lengths[2]}")
print(f"Length range: {sequence_lengths[3]} - {sequence_lengths[4]} timesteps")
print(f"Quartiles: Q1={sequence_lengths[5]}, Q2={sequence_lengths[6]}, Q3={sequence_lengths[7]}")
print(f"95th percentile: {sequence_lengths[8]} timesteps")

# 50Hz前提での時間計算
print(f"\n⏱️ Time Duration (assuming 50Hz sampling):")
print(f"  • Average sequence duration: {sequence_lengths[1]/50:.2f} seconds")
print(f"  • Median sequence duration: {sequence_lengths[6]/50:.2f} seconds")
print(f"  • Longest sequence duration: {sequence_lengths[4]/50:.1f} seconds")

# ジェスチャー別のシーケンス長分析
print("\n📊 Sequence Length by Gesture Type:")
gesture_lengths = conn.execute("""
    SELECT 
        gesture,
        COUNT(DISTINCT sequence_id) as num_sequences,
        ROUND(AVG(length), 1) as avg_length,
        ROUND(MIN(length), 1) as min_length,
        ROUND(MAX(length), 1) as max_length
    FROM (
        SELECT 
            sequence_id, 
            gesture,
            COUNT(*) as length
        FROM "cmi_detect_behavior_with_sensor_data".train
        GROUP BY sequence_id, gesture
    )
    GROUP BY gesture
    ORDER BY avg_length DESC
    LIMIT 10
""").fetchdf()

print(gesture_lengths.to_string(index=False))

# 参加者別のシーケンス数分析
print("\n👥 Sequence Count per Participant:")
participant_sequences = conn.execute("""
    SELECT 
        COUNT(DISTINCT subject) as total_participants,
        ROUND(AVG(seq_count), 1) as avg_sequences_per_participant,
        MIN(seq_count) as min_sequences,
        MAX(seq_count) as max_sequences,
        ROUND(STDDEV(seq_count), 1) as std_sequences
    FROM (
        SELECT 
            subject, 
            COUNT(DISTINCT sequence_id) as seq_count
        FROM "cmi_detect_behavior_with_sensor_data".train
        GROUP BY subject
    )
""").fetchone()

print(f"Participants: {participant_sequences[0]}")
print(f"Sequences per participant: mean={participant_sequences[1]}, std={participant_sequences[4]}")
print(f"Sequence range per participant: {participant_sequences[2]} - {participant_sequences[3]}")

# 時系列の連続性チェック
print("\n🔗 Sequence Continuity Check:")
continuity_check = conn.execute("""
    SELECT 
        COUNT(*) as total_gaps,
        AVG(gap_size) as avg_gap_size,
        MAX(gap_size) as max_gap_size
    FROM (
        SELECT 
            sequence_id,
            sequence_counter - LAG(sequence_counter) OVER (PARTITION BY sequence_id ORDER BY sequence_counter) as gap_size
        FROM "cmi_detect_behavior_with_sensor_data".train
    )
    WHERE gap_size > 1
""").fetchone()

if continuity_check[0] > 0:
    print(f"⚠️  Found {continuity_check[0]} gaps in sequences")
    print(f"   Average gap size: {continuity_check[1]:.1f}")
    print(f"   Maximum gap size: {continuity_check[2]}")
else:
    print("✅ All sequences are continuous (no missing timesteps)")

print("\n" + "=" * 60)

In [ ]:
# 3. 🎯 ターゲット変数の詳細分析
print("🎯 3. TARGET VARIABLE CORRELATION & CLASS BALANCE ANALYSIS")
print("=" * 60)

# ジェスチャーとsequence_typeの関係
print("🔄 Gesture vs Sequence Type Relationship:")
gesture_sequence_type = conn.execute("""
    SELECT 
        gesture,
        sequence_type,
        COUNT(*) as count,
        ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER(), 2) as percentage
    FROM "cmi_detect_behavior_with_sensor_data".train
    GROUP BY gesture, sequence_type
    ORDER BY count DESC
    LIMIT 15
""").fetchdf()

print(gesture_sequence_type.to_string(index=False))

# 行動フェーズの詳細分析
print("\n📊 Behavior Phase Distribution by Gesture (Top 5):")
behavior_phase = conn.execute("""
    SELECT 
        gesture,
        behavior,
        COUNT(*) as count,
        ROUND(AVG(sequence_counter), 1) as avg_timestamp
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE gesture IN (
        SELECT gesture 
        FROM "cmi_detect_behavior_with_sensor_data".train 
        GROUP BY gesture 
        ORDER BY COUNT(*) DESC 
        LIMIT 5
    )
    GROUP BY gesture, behavior
    ORDER BY gesture, count DESC
""").fetchdf()

print(behavior_phase.to_string(index=False))

# クラス不均衡の詳細評価
print("\n⚖️ Class Imbalance Analysis:")

# Binary classification (sequence_type)
binary_balance = conn.execute("""
    SELECT 
        sequence_type,
        COUNT(*) as count,
        ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train), 2) as percentage
    FROM "cmi_detect_behavior_with_sensor_data".train
    GROUP BY sequence_type
    ORDER BY count DESC
""").fetchdf()

print("Binary Classification (sequence_type):")
print(binary_balance.to_string(index=False))

# Multi-class gesture distribution
gesture_balance = conn.execute("""
    SELECT 
        gesture,
        COUNT(*) as count,
        ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train), 2) as percentage,
        COUNT(DISTINCT subject) as participants
    FROM "cmi_detect_behavior_with_sensor_data".train
    GROUP BY gesture
    ORDER BY count DESC
""").fetchdf()

print(f"\nMulti-class Gesture Distribution (18 classes):")
print(gesture_balance.to_string(index=False))

# 不均衡比率の計算
max_class = gesture_balance['count'].max()
min_class = gesture_balance['count'].min()
imbalance_ratio = max_class / min_class

print(f"\n📈 Class Imbalance Metrics:")
print(f"  • Most frequent gesture: {gesture_balance.iloc[0]['gesture']} ({gesture_balance.iloc[0]['count']:,} samples)")
print(f"  • Least frequent gesture: {gesture_balance.iloc[-1]['gesture']} ({gesture_balance.iloc[-1]['count']:,} samples)")
print(f"  • Imbalance ratio: {imbalance_ratio:.1f}:1")
print(f"  • Macro F1 challenge level: {'HIGH' if imbalance_ratio > 10 else 'MEDIUM' if imbalance_ratio > 3 else 'LOW'}")

print("\n" + "=" * 60)

In [ ]:
# 2. 🔍 センサーデータ分布分析
print("🔍 2. SENSOR DATA DISTRIBUTION ANALYSIS")
print("=" * 60)

# IMUセンサーの基本統計量
print("🎯 IMU Sensor Statistics:")
imu_stats = conn.execute("""
    SELECT 
        'acc_x' as sensor,
        COUNT(*) - COUNT(acc_x) as null_count,
        ROUND(AVG(acc_x), 4) as mean_val,
        ROUND(STDDEV(acc_x), 4) as std_val,
        ROUND(MIN(acc_x), 4) as min_val,
        ROUND(MAX(acc_x), 4) as max_val,
        ROUND(PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY acc_x), 4) as median_val
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE acc_x IS NOT NULL
    
    UNION ALL
    
    SELECT 
        'acc_y',
        COUNT(*) - COUNT(acc_y),
        ROUND(AVG(acc_y), 4),
        ROUND(STDDEV(acc_y), 4),
        ROUND(MIN(acc_y), 4),
        ROUND(MAX(acc_y), 4),
        ROUND(PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY acc_y), 4)
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE acc_y IS NOT NULL
    
    UNION ALL
    
    SELECT 
        'acc_z',
        COUNT(*) - COUNT(acc_z),
        ROUND(AVG(acc_z), 4),
        ROUND(STDDEV(acc_z), 4),
        ROUND(MIN(acc_z), 4),
        ROUND(MAX(acc_z), 4),
        ROUND(PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY acc_z), 4)
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE acc_z IS NOT NULL
""").fetchdf()

print(imu_stats.to_string(index=False))

# 回転センサー（クォータニオン）の統計量
print("\n🔄 Rotation Quaternion Statistics:")
rot_stats = conn.execute("""
    SELECT 
        'rot_w' as sensor,
        COUNT(*) - COUNT(rot_w) as null_count,
        ROUND(AVG(rot_w), 4) as mean_val,
        ROUND(STDDEV(rot_w), 4) as std_val,
        ROUND(MIN(rot_w), 4) as min_val,
        ROUND(MAX(rot_w), 4) as max_val
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE rot_w IS NOT NULL
    
    UNION ALL
    
    SELECT 'rot_x', COUNT(*) - COUNT(rot_x), ROUND(AVG(rot_x), 4), ROUND(STDDEV(rot_x), 4), ROUND(MIN(rot_x), 4), ROUND(MAX(rot_x), 4)
    FROM "cmi_detect_behavior_with_sensor_data".train WHERE rot_x IS NOT NULL
    
    UNION ALL
    
    SELECT 'rot_y', COUNT(*) - COUNT(rot_y), ROUND(AVG(rot_y), 4), ROUND(STDDEV(rot_y), 4), ROUND(MIN(rot_y), 4), ROUND(MAX(rot_y), 4)
    FROM "cmi_detect_behavior_with_sensor_data".train WHERE rot_y IS NOT NULL
    
    UNION ALL
    
    SELECT 'rot_z', COUNT(*) - COUNT(rot_z), ROUND(AVG(rot_z), 4), ROUND(STDDEV(rot_z), 4), ROUND(MIN(rot_z), 4), ROUND(MAX(rot_z), 4)
    FROM "cmi_detect_behavior_with_sensor_data".train WHERE rot_z IS NOT NULL
""").fetchdf()

print(rot_stats.to_string(index=False))

# 温度センサーの統計量
print("\n🌡️ Thermopile Sensor Statistics:")
thm_stats = conn.execute("""
    SELECT 
        'thm_1' as sensor,
        COUNT(*) - COUNT(thm_1) as null_count,
        ROUND(AVG(thm_1), 2) as mean_val,
        ROUND(STDDEV(thm_1), 2) as std_val,
        ROUND(MIN(thm_1), 2) as min_val,
        ROUND(MAX(thm_1), 2) as max_val
    FROM "cmi_detect_behavior_with_sensor_data".train
    WHERE thm_1 IS NOT NULL
    
    UNION ALL SELECT 'thm_2', COUNT(*) - COUNT(thm_2), ROUND(AVG(thm_2), 2), ROUND(STDDEV(thm_2), 2), ROUND(MIN(thm_2), 2), ROUND(MAX(thm_2), 2)
    FROM "cmi_detect_behavior_with_sensor_data".train WHERE thm_2 IS NOT NULL
    UNION ALL SELECT 'thm_3', COUNT(*) - COUNT(thm_3), ROUND(AVG(thm_3), 2), ROUND(STDDEV(thm_3), 2), ROUND(MIN(thm_3), 2), ROUND(MAX(thm_3), 2)
    FROM "cmi_detect_behavior_with_sensor_data".train WHERE thm_3 IS NOT NULL
    UNION ALL SELECT 'thm_4', COUNT(*) - COUNT(thm_4), ROUND(AVG(thm_4), 2), ROUND(STDDEV(thm_4), 2), ROUND(MIN(thm_4), 2), ROUND(MAX(thm_4), 2)
    FROM "cmi_detect_behavior_with_sensor_data".train WHERE thm_4 IS NOT NULL
    UNION ALL SELECT 'thm_5', COUNT(*) - COUNT(thm_5), ROUND(AVG(thm_5), 2), ROUND(STDDEV(thm_5), 2), ROUND(MIN(thm_5), 2), ROUND(MAX(thm_5), 2)
    FROM "cmi_detect_behavior_with_sensor_data".train WHERE thm_5 IS NOT NULL
""").fetchdf()

print(thm_stats.to_string(index=False))

print("\n" + "=" * 60)

In [ ]:
# 1. 📊 包括的データプロファイリング
print("🔍 1. COMPREHENSIVE DATA PROFILING")
print("=" * 60)

# 基本データ構造の確認
print("📋 Dataset Structure:")
structure_info = conn.execute("""
    SELECT 
        'train' as table_name,
        COUNT(*) as total_rows,
        COUNT(DISTINCT subject) as unique_participants,
        COUNT(DISTINCT sequence_id) as unique_sequences,
        MIN(sequence_counter) as min_counter,
        MAX(sequence_counter) as max_counter
    FROM "cmi_detect_behavior_with_sensor_data".train
    
    UNION ALL
    
    SELECT 
        'test' as table_name,
        COUNT(*) as total_rows,
        COUNT(DISTINCT subject) as unique_participants,
        COUNT(DISTINCT sequence_id) as unique_sequences,
        MIN(sequence_counter) as min_counter,
        MAX(sequence_counter) as max_counter
    FROM "cmi_detect_behavior_with_sensor_data".test
""").fetchdf()

print(structure_info.to_string(index=False))

# データ型と非NULL値の詳細確認
print("\n🧮 Column Data Types and Non-null Counts:")
column_info = conn.execute("""
    SELECT 
        column_name,
        data_type,
        is_nullable
    FROM information_schema.columns 
    WHERE table_schema = 'cmi_detect_behavior_with_sensor_data' 
    AND table_name = 'train'
    ORDER BY ordinal_position
""").fetchdf()

print(f"Total columns: {len(column_info)}")

# カテゴリ別の列数
categorical_cols = ['row_id', 'sequence_type', 'sequence_id', 'subject', 'orientation', 'behavior', 'phase', 'gesture']
sensor_cols = [col for col in column_info['column_name'] if col not in categorical_cols]

print(f"  📝 Categorical columns: {len(categorical_cols)}")
print(f"  🔢 Sensor columns: {len(sensor_cols)}")

# センサー別分類
imu_cols = [col for col in sensor_cols if col.startswith(('acc_', 'rot_'))]
thermopile_cols = [col for col in sensor_cols if col.startswith('thm_')]
tof_cols = [col for col in sensor_cols if col.startswith('tof_')]

print(f"    🎯 IMU sensors: {len(imu_cols)} ({imu_cols})")
print(f"    🌡️  Thermopile sensors: {len(thermopile_cols)} ({thermopile_cols})")
print(f"    📡 ToF sensors: {len(tof_cols)} (5 sensors × 64 channels)")

print("\n" + "=" * 60)

# 🧠 CMI BFRB Detection - 本格的EDA

## 📋 分析プラン

### 🎯 目標
- **コンペ概要**: Body-Focused Repetitive Behaviors (BFRB) 検出
- **評価指標**: 0.5×(Binary F1 + Macro F1) 
- **データ**: マルチモーダルセンサー時系列データ (50Hz)
- **参加者**: 81名の訓練データ、18種類のジェスチャー

### 📊 詳細分析項目
1. **データプロファイリング** - 基本統計量と品質評価
2. **センサーデータ分析** - 分布、外れ値、ノイズ特性
3. **ターゲット変数分析** - クラス不均衡、相関関係
4. **時系列パターン分析** - シーケンス構造、周期性
5. **欠損値パターン** - センサー故障、データ品質
6. **参加者特性分析** - 人口統計学的特徴
7. **特徴相関分析** - マルチコリニアリティ、特徴選択
8. **マルチモーダル分析** - センサー融合の可能性
9. **CV戦略検証** - データリーク防止策
10. **特徴工学提案** - ドメイン知識活用

---

In [5]:
# Debug database tables issue
print("=== DEBUG DATABASE TABLES ===")

# Check if connection exists
try:
    print(f"Connection status: {conn}")
except NameError:
    print("Connection not found, creating new one...")
    import duckdb
    conn = duckdb.connect('/home/wsl/dev/my-study/ml/ml-note/kaggle_datasets.duckdb')

# Check all schemas
print("\n=== ALL SCHEMAS ===")
schemas = conn.execute("SELECT schema_name FROM information_schema.schemata").fetchall()
for schema in schemas:
    print(f"  - {schema[0]}")

# Check all tables with schema
print("\n=== ALL TABLES WITH SCHEMA ===")
tables = conn.execute("SELECT table_schema, table_name FROM information_schema.tables WHERE table_type = 'BASE TABLE'").fetchall()
print(f"Found {len(tables)} tables:")
for schema, table in tables:
    print(f"  - {schema}.{table}")

# Try SHOW TABLES in different schemas
print("\n=== SHOW TABLES ===")
try:
    show_tables = conn.execute("SHOW TABLES").fetchall()
    print(f"Default schema tables: {len(show_tables)} found")
    for table in show_tables:
        print(f"  - {table[0]}")
except Exception as e:
    print(f"Error with SHOW TABLES: {e}")

# Try to access the CMI tables directly
print("\n=== TESTING CMI SCHEMA ACCESS ===")
try:
    result = conn.execute('SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train').fetchone()
    print(f"✅ cmi_detect_behavior_with_sensor_data.train accessible: {result[0]} rows")
except Exception as e:
    print(f"❌ Error accessing cmi_detect_behavior_with_sensor_data.train: {e}")

# List all objects in kaggle_datasets schema
print("\n=== KAGGLE_DATASETS SCHEMA CONTENTS ===")
try:
    result = conn.execute('SHOW TABLES FROM "kaggle_datasets"."cmi_detect_behavior_with_sensor_data"').fetchall()
    print(f"Tables in cmi_detect_behavior_with_sensor_data schema:")
    for table in result:
        print(f"  - {table[0]}")
except Exception as e:
    print(f"Error listing cmi schema tables: {e}")

print("\n=== DATABASE FILE INFO ===")
import os
db_path = '/home/wsl/dev/my-study/ml/ml-note/kaggle_datasets.duckdb'
if os.path.exists(db_path):
    size = os.path.getsize(db_path)
    print(f"✅ Database file exists, size: {size:,} bytes ({size/1024/1024:.1f} MB)")
else:
    print("❌ Database file does not exist!")

=== DEBUG DATABASE TABLES ===
Connection status: <duckdb.duckdb.DuckDBPyConnection object at 0x7f69477016b0>

=== ALL SCHEMAS ===
  - cmi_detect_behavior_with_sensor_data
  - main
  - playground_series_s5e7
  - information_schema
  - main
  - pg_catalog
  - main

=== ALL TABLES WITH SCHEMA ===
Found 7 tables:
  - cmi_detect_behavior_with_sensor_data.test
  - cmi_detect_behavior_with_sensor_data.test_demographics
  - cmi_detect_behavior_with_sensor_data.train
  - cmi_detect_behavior_with_sensor_data.train_demographics
  - playground_series_s5e7.sample_submission
  - playground_series_s5e7.test
  - playground_series_s5e7.train

=== SHOW TABLES ===
Default schema tables: 0 found

=== TESTING CMI SCHEMA ACCESS ===
✅ cmi_detect_behavior_with_sensor_data.train accessible: 574945 rows

=== KAGGLE_DATASETS SCHEMA CONTENTS ===
Error listing cmi schema tables: Parser Error: syntax error at or near "FROM"

=== DATABASE FILE INFO ===
✅ Database file exists, size: 286,273,536 bytes (273.0 MB)


In [ ]:
import duckdb
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("🔍 CMI BFRB Detection Dataset - EDA")
print("=" * 50)

# Connect to the DuckDB database
conn = duckdb.connect('/home/wsl/dev/my-study/ml/ml-note/kaggle_datasets.duckdb')

# Show tables from the correct schema
print("📊 Available tables in cmi_detect_behavior_with_sensor_data schema:")
tables = conn.execute("""
    SELECT table_name 
    FROM information_schema.tables 
    WHERE table_schema = 'cmi_detect_behavior_with_sensor_data'
    ORDER BY table_name
""").fetchall()

for table in tables:
    print(f"  ✅ {table[0]}")

print(f"\n📈 Database size: {286.3:.1f} MB with {len(tables)} tables")
print(f"🎯 Target dataset: CMI Body-Focused Repetitive Behaviors Detection")
print("=" * 50)

## Key Dataset Findings

### Data Size
- **Train**: 574,945 rows across 81 participants (8,151 sequences)
- **Test**: 107 rows across 2 participants (2 sequences)  
- **No participant overlap** between train and test sets

### Data Distribution
- Average ~7,098 rows per participant in training
- Average ~71 sequences per participant in training
- Test set appears to be a small sample for submission format

### Important Notes
- This is a **time series** dataset with sequence structure
- Need **GroupKFold by participant** to prevent data leakage
- Large training set (~575k timesteps) suggests 50Hz sampling over multiple sessions

In [6]:
# Re-establish database connection for this cell
import duckdb
conn = duckdb.connect('/home/wsl/dev/my-study/ml/ml-note/kaggle_datasets.duckdb')

# Data size and distribution analysis
print("=== DATA SIZE ANALYSIS ===")

# Check row counts
train_count = conn.execute('SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train').fetchone()[0]
test_count = conn.execute('SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".test').fetchone()[0]
train_demo_count = conn.execute('SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".train_demographics').fetchone()[0]
test_demo_count = conn.execute('SELECT COUNT(*) FROM "cmi_detect_behavior_with_sensor_data".test_demographics').fetchone()[0]

print(f"Train data rows: {train_count:,}")
print(f"Test data rows: {test_count:,}")
print(f"Train demographics: {train_demo_count:,}")
print(f"Test demographics: {test_demo_count:,}")

print("\n=== PARTICIPANT ANALYSIS ===")

# Unique participants
train_participants = conn.execute('SELECT COUNT(DISTINCT subject) FROM "cmi_detect_behavior_with_sensor_data".train').fetchone()[0]
test_participants = conn.execute('SELECT COUNT(DISTINCT subject) FROM "cmi_detect_behavior_with_sensor_data".test').fetchone()[0]

print(f"Unique participants in train: {train_participants}")
print(f"Unique participants in test: {test_participants}")

# Check for overlap in participants between train and test
overlap_check = conn.execute('''
    SELECT COUNT(*) FROM (
        SELECT subject FROM "cmi_detect_behavior_with_sensor_data".train 
        INTERSECT 
        SELECT subject FROM "cmi_detect_behavior_with_sensor_data".test
    )
''').fetchone()[0]

print(f"Participants appearing in both train and test: {overlap_check}")

print("\n=== SEQUENCE ANALYSIS ===")

# Sequence counts
train_sequences = conn.execute('SELECT COUNT(DISTINCT sequence_id) FROM "cmi_detect_behavior_with_sensor_data".train').fetchone()[0]
test_sequences = conn.execute('SELECT COUNT(DISTINCT sequence_id) FROM "cmi_detect_behavior_with_sensor_data".test').fetchone()[0]

print(f"Unique sequences in train: {train_sequences}")
print(f"Unique sequences in test: {test_sequences}")

=== DATA SIZE ANALYSIS ===
Train data rows: 574,945
Test data rows: 107
Train demographics: 81
Test demographics: 2

=== PARTICIPANT ANALYSIS ===
Unique participants in train: 81
Unique participants in test: 2
Participants appearing in both train and test: 0

=== SEQUENCE ANALYSIS ===
Unique sequences in train: 8151
Unique sequences in test: 2


# CMI BFRB Detection - Exploratory Data Analysis

## Dataset Schema Analysis

### Data Structure Overview
- **Train table**: Contains sensor data with target labels (behavior, gesture, phase)
- **Test table**: Contains sensor data without target labels  
- **Demographics tables**: Participant information (age, sex, handedness, physical measurements)

### Sensor Features
- **IMU sensors**: acc_x/y/z (accelerometer), rot_w/x/y/z (rotation quaternion) = 7 features
- **Thermopile sensors**: thm_1 to thm_5 = 5 features  
- **ToF sensors**: tof_1 to tof_5 with 64 values each (v0 to v63) = 320 features
- **Total sensor features**: 332 per timestep

### Target Variables (Train only)
- **behavior**: Binary classification (BFRB vs non-BFRB)
- **gesture**: Multi-class classification (specific gesture types)
- **phase**: Multi-class classification (gesture phases)